api/C/include/csi_dl_uart.h

```diff
@@ -0,0 +1,356 @@
+    UART_EVENT_RECEIVE_COMPLETE, /* Data reception complete */
+    UART_EVENT_RECEIVE_FIFO_READABLE, /* Data available in UART Rx FIFO */
+    UART_EVENT_BREAK_INTR, /* Data reception interrupted (break signal from sender) */
+    UART_EVENT_ERROR_OVERFLOW, /* Data receive FIFO overflowed */
+    UART_EVENT_ERROR_PARITY, /* Data reception parity error */
+    UART_EVENT_ERROR_FRAMING, /* Framing error (no effective stop bit) */
+    UART_EVENT_TX_ABORT, /* Transmission aborted by calling csi_uart_stop_tx */
+    UART_EVENT_RX_ABORT, /* Reception aborted by calling csi_uart_stop_rx */
+} csi_uart_event_t;
+
+/*
+ * Function prototype for user's callback function to be run by the UART driver on
+ * UART events.
+ *
+ * @param callback_ctx: Context pointer that was passed into
+ * csi_uart_register_callback.
+ * @param event: Event giving rise to the callback.
+ * @param bytes_transferred: number of bytes of transfer completed up to this
+ * point.
+ */
+typedef void (csi_uart_callback_t)(void *callback_ctx, csi_uart_event_t event, unsigned bytes_transferred);
+
+/*
+ * Structure of information concerning UART state
+ */
+typedef struct {
+    bool readable;
+    bool writable;
+    bool overrun_error;
+    bool parity_error;
+    bool framing_error;
+    bool tx_fifo_empty;
+} csi_uart_state_t;
+
+
+/*
+ * Initialize a UART driver instance and associate it with a physical channel.
+ * This function must be called before calling any other functions in this module.
+ *
+ * @param uart: Context structure for this driver instance: initialized by this
+ * function call.  The structure declaration csi_uart_t is published by the BSP in
+ * csi_hl_bsp_interrupts.h but should be considered private to the BSP.
+ * Application code instantiates this structure and passes in a pointer to it.
+ * @param idx: Index of the UART, starting at 0 for the first-numbered device in
+ * the system.  The BSP may publish macros giving meaningful names to UART
+ * instances.
+ * @param irq_reg_fn: Pointer to a function that the BSP code can use to register
+ * an interrupt handler to support asynchronous callbacks.  May be NULL in which
+ * case asynchronous UART functions are not supported.
+ * @param irq_system_ctx: Context pointer for the interrupt subsystem, to pass into
+ * irq_reg_fn. May be NULL in which case asynchronous UART functions are not
+ * supported.
+ * @param source: Enumeration of the interrupt source, to be passed into
+ * irq_reg_fn.  The user must select the appropriate value to select the interrupt
+ * associated with the UART being initialized. Unused if irq_reg_fn and / or
+ * irq_system_ctx are NULL.
+ * @param wait_fn: Pointer to a function that the BSP should call to implement a
+ * wait while polling.
+ * @return : Status of initialisation operation
+ */
+csi_status_t csi_uart_init(csi_uart_t *uart, unsigned idx, csi_register_isr_wrapper_t *irq_reg_fn, void *irq_system_ctx, int source, csi_poll_wait_wrapper_t *wait_fn);
```

Comment:
Will this `uart` struct already contains the information of the uart object, i think `idx` may not be needed.

Reply:
No, the uart structure is private to the BSP.  The application code instantiates an uninitialized structure and passes a pointer to it, then the structure contents are populated by this init function.  The contents of the structure are not defined by our API.  So the application code needs to inform the BSP of the UART index through the idx parameter.

Reply:
Ok, then it make sense to have a idx passed